File: scalalib/test/resources/classloader-test/test/src/ClassLoaderTests.scala
package mill.scalalib

import utest._

object ClassLoaderTests extends TestSuite {
  try {
    getClass().getClassLoader().loadClass("com.sun.nio.zipfs.ZipFileSystemProvider")
  } catch {
    case _: ClassNotFoundException if !System.getProperty("java.specification.version").startsWith("1.") =>
      // Don't fail on Java 9+
  }
  val tests = Tests {}
}

File: libs/javalib/test/resources/classloader-test/test/src/ClassLoaderTests.java
package mill.javalib;

import org.junit.Test;
import static org.junit.Assert.*;

public class ClassLoaderTests {
    @Test
    public void comSunClassesExistInTestsClasspath() throws Exception {
        String javaVersion = System.getProperty("java.specification.version");
        boolean isJava8 = javaVersion.startsWith("1.");
        if (isJava8) {
            // This should not throw ClassNotFoundException on Java 8
            getClass().getClassLoader().loadClass("com.sun.nio.zipfs.ZipFileSystemProvider");
        }
        // On newer Java versions, this test passes trivially
        assertTrue(true);
    }
}
